brave/src/main/java/com/linecorp/armeria/common/brave/RequestContextCurrentTraceContext.java

```diff
@@ -0,0 +1,311 @@
+    private static final ThreadLocal<TraceContext> THREAD_LOCAL_CONTEXT = new ThreadLocal<>();
+
+    private static final Scope INCOMPLETE_CONFIGURATION_SCOPE = new Scope() {
+        @Override
+        public void close() {
+        }
+
+        @Override
+        public String toString() {
+            return "IncompleteConfigurationScope";
+        }
+    };
+
+    private static final Scope INITIAL_REQUEST_SCOPE = new Scope() {
+        @Override
+        public void close() {
+            // Don't remove the outer-most context (client or server request)
+        }
+
+        @Override
+        public String toString() {
+            return "InitialRequestScope";
+        }
+    };
+
+    static final class Builder extends CurrentTraceContext.Builder {
+
+        @Override
+        public CurrentTraceContext build() {
+            return new RequestContextCurrentTraceContext(this);
+        }
+    }
+
+    /**
+     * Use this when you need customizations such as log integration via
+     * {@linkplain Builder#addScopeDecorator(ScopeDecorator)}.
+     *
+     * @see Tracing.Builder#currentTraceContext(CurrentTraceContext)
+     */
+    public static CurrentTraceContext.Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Ensures the specified {@link Tracing} uses a {@link RequestContextCurrentTraceContext}.
+     *
+     * @throws IllegalStateException if {@code tracing} does not use {@link RequestContextCurrentTraceContext}
+     */
+    public static void ensureScopeUsesRequestContext(Tracing tracing) {
+        requireNonNull(tracing, "tracing");
+        final PingPongExtra extra = new PingPongExtra();
+        // trace contexts are not recorded until Tracer.toSpan, so this won't end up as junk data
+        final TraceContext dummyContext = TraceContext.newBuilder().traceId(1).spanId(1)
+                                                      .extra(Collections.singletonList(extra)).build();
+        final boolean scopeUsesRequestContext;
+        try (Scope scope = tracing.currentTraceContext().newScope(dummyContext)) {
+            scopeUsesRequestContext = extra.isPong();
+        }
+        if (!scopeUsesRequestContext) {
+            throw new IllegalStateException(
+                    "Tracing.currentTraceContext is not a " + RequestContextCurrentTraceContext.class
+                            .getSimpleName() + " scope. " +
+                    "Please call Tracing.Builder.currentTraceContext(" + RequestContextCurrentTraceContext.class
+                            .getSimpleName() + ".DEFAULT).");
+        }
+    }
+
+    /**
+     * Use this to ensure the trace context propagates to children.
+     *
+     * <p>Ex.
+     * <pre>{@code
+     *  // Ensure the trace context propagates to children
+     * ctx.onChild(RequestContextCurrentTraceContext::copy);
+     * }</pre>
+     */
+    public static void copy(RequestContext src, RequestContext dst) {
```

Comment:
Should a user ever need to invoke this?

Reply:
This is used in `BraveClient` and `BraveService` so that this should be public. Do you want me to this into the internal package?

Reply:
Ah then yeah think it should be in internal package.